### DISCUSSION TITLE
ontop option in Image?

### DISCUSSION THREAD

**Discussion by baba-yaga** [marcomusy/vedo#1001] [Q&A]
Happy New Year, everyone!
I discovered that there is no longer the option ontop=True in Image() function which was very useful to me in Image2D() in the previous versions of vedo. How can I keep a picture on top of everything else in a renderer now?
Thank you!

**Comment by marcomusy** [comment]
Happy new Year Sergei! 
Yes try e.g.:
```py
from vedo import *
s = Sphere().lw(1)
img = Image(dataurl+"images/dog.jpg") # this is 3D
img = img.clone2d("top-right").ontop(True) # this is a 2D copy of it
show(s, img, bg='wheat')
```

Also check out [this example](https://github.com/marcomusy/vedo/blob/master/examples/pyplot/embed_matplotlib.py).

**Reply by baba-yaga** [reply]
Happy New Year, Marco! 
Thank you - it's almost what I need. The problem is that the image and it's 3d embedding show differently.
```
from vedo import *
img = Image(dataurl + 'images/dog.jpg')
img2d = img.clone2d()

show(img.alpha(0.5), img2d.alpha(0.5)).close()

``` 
I cannot get rid of the white space in 3d image, even if I explicitly set the renderer's size. 
Well, I can still make them look the same, but I have a strong feeling that my way is an overkill )

```
# Get and save the camera used for 2d-image
plt = Plotter(size=img2d.shape)
plt.show(img2d, interactive=False)
cam = plt.camera
close()

plt = Plotter(size=img2d.shape, axes=1)
plt += img2d.alpha(0.5)
# shift 3d-image so that it is not anchored in the centre of the frame
plt += img.shift((-0.5 * img.shape[0], -0.5 * img.shape[1], 0)).alpha(0.5)
plt.show(camera=cam).close()  # re-using the camera

```

**Reply by marcomusy** [reply]
Hi, I'm not completely sure of what you are trying to achieve :) what about this
```py
from vedo import *

img = Image(dataurl + 'images/dog.jpg')

#############################################
plt = Plotter(size=img.shape).parallel_projection(True)
plt += img.alpha(0.5)
# plt.reset_camera(tight=0.0) # same as "zoom="tightest"
plt.show(zoom="tightest").close()  

#OR #########################################
img2d = img.clone2d()
plt = Plotter(size=img2d.shape)
plt += img2d
plt.show().close()
```

**Reply by baba-yaga** [reply]
What I am trying to achieve is to have a 2D image shown on top of a 3D mesh, but I see now that clone2d result is basically just a background which stays fixed independently of the manipulations in viewport. To illustrate what I want, imagine you want to use an image as a stencil to texture a 3D mesh. For instance, I want a dog to appear on a face of a cube:

```
from vedo import *
from vtkmodules.vtkRenderingCore import vtkCamera

img = Image(dataurl+"images/dog.jpg")  # 3D embedding of the image
sz = img.shape

cp = (sz[0]/2, sz[1]/2, 0.7 * sz[1])
fp = (sz[0]/2, sz[1]/2, 0)

cam1 = vtkCamera()
cam1.SetPosition(pos)
cam1.SetFocalPoint(fp)
cam1.SetViewAngle(90)

C = Cube(fp, 300, c='yellow').clean().lighting("off")

```
If I just plot them together, a part of the image is covered by the cube

```
plt = Plotter(size=sz)
plt += img, C
plt.show(camera=cam1)

```
<img width="287" alt="Screenshot 2024-01-09 at 23 53 43" src="https://github.com/marcomusy/vedo/assets/2671379/db1747d9-03a1-47a7-9cb4-77483a815f26">

To make it show through, either I need to move back the cube (by some reason, I do not want to do it), or to move the image forward scaling it so that it looks the same. So I do

```
def make_stencil(image3d: Image, vtkcam, mesh: Mesh):
    cp2 = vtkcam.GetPosition()[2]
    zmin = np.array([cp2-v[2] for v in mesh.vertices]).min()  # distance from cam to mesh along z-axis
    scale = zmin / vtkcam.GetDistance()
    T = LinearTransform().scale(scale, origin=cp)
    return image3d.apply_transform(T)

st = make_stencil(img.clone(), cam1, C)

plt = Plotter()
plt += C, st.alpha(0.5)
plt += Point(cp, r=20)
plt.show(camera=cam1, axes=1).close()

```
<img width="376" alt="Screenshot 2024-01-10 at 00 01 03" src="https://github.com/marcomusy/vedo/assets/2671379/7329bbfe-c4f1-4d47-93ac-a65ff89f9df5">

So now the image is seen everywhere which allows me to transfer the texture with a defined brush at precise locations not blindly. What I still do not know how to achieve and would very much liked to have your advice. If I work around the corners, I do not want the texture to spill on another face. So I would like to be able to zoom in the view. With your previous help,

```
plt = Plotter()
plt += C, st.alpha(0.5)
plt += Point(cp, r=20)
plt.show(camera=cam1, axes=1, mode=MousePan()).close()

```
allows to zoom in without rotation, but the zoom is about the centre of the renderer window (changing the view angle of the camera, as I understand). Pan, however, destroys the alignment. What I would like to have is a zoom about the location of the mouse so that moving the mouse close to a corner and zooming in does not destroys the alignment of the mesh and the stencil. Alternatively, to have another sort of pan which moves the screen (together with the coordinate system) and not the camera. Then moving the scene so that the corner is about the centre of the frame, the usual camera zoom could be used to preserve the alignment. Is that possible?

**Reply by baba-yaga** [reply]
You may wonder, why not to simply make the cube half-transparent? The reason is that my `cube' is a human face mesh which has quite a complicated structure. Making it self-transparent would reveal the tongue, teeth, nostrils, back side of the head, making the view absolutely messy.

**Reply by baba-yaga** [reply]
PS. I do not like that my function make_stencil modifies its argument: image3d. I tried to clone the argument inside the function, but it does not work if there is already an transformation applied to it. The problem is discussed here: [https://github.com/marcomusy/vedo/discussions/1011](https://github.com/marcomusy/vedo/discussions/1011)

**Reply by marcomusy** [reply]
I think what you are looking for is **texturing a mesh** (?)
If so you need to define you (u,v) texture coordinates, then simply:

```py
from vedo import *
img = Image(dataurl+"images/dog.jpg")
cube = Cube()
cube.texture(img, tcoords=None) # or simply cube.texture("dog.jpg")
show(cube, axes=1).close() 
```
![Screenshot from 2024-01-10 15-50-19](https://github.com/marcomusy/vedo/assets/32848391/4f433093-80ed-4bf7-8876-7f93c1ad5750)

**Reply by baba-yaga** [reply]
Well, it would be a too easy task just to texture the cube ) Actually, the primary task is photogrammetry: restore 3D object from images shot from different angles. The first step is alignment of the 3D model to the image using a small set of positioning landmarks. Next comes adjusting the other landmarks by moving them in the focal plane and warping the model. And here I need a precision with ability to zoom in without breaking the alignment of the image and the 3D model. Here how it looks in real:
![Screenshot 2023-12-19 at 19 20 23](https://github.com/marcomusy/vedo/assets/2671379/4bc1d3b2-0131-4deb-a97a-a2e167c9d6c8)

You see a grey 3D model coming out of the image which cuts through it. So I need to move the image forward to make it visible at all places and then adjust the projections of the model vertices to fit the image. If I change the opacity of the model, I get a mess of all inner parts (teeth, nostrils, etc.) showing too
<img width="346" alt="Screenshot 2024-01-10 at 16 18 08" src="https://github.com/marcomusy/vedo/assets/2671379/ae0f4e79-e9b0-4257-9713-90c080dbcc2f">

**Reply by baba-yaga** [reply]
Then texturing is also a complicated thing. I do not want to transfer the texture on the eyeballs, for instance, so again precision is needed. The same type as zooming the whole computer screen

**Reply by marcomusy** [reply]
OK - to be honest i'm not familiar at all with the subject of texturing a polygonal mesh :)

My suggestion (which may be completely stupid or naive)
- you can create an image and make a mesh out of it with `img.tomesh()`
- assign to this mesh x,y arrays by normalizing it the way you like (so this would serve as (u,v) later)
- write a callback function that allows you to click and select landmarks
- snap the 3d image mesh to some landmarks
- in the same callback you can reset/slide along an axis the position of the image wrt the model
- read out the values of the original [u,v] onto the mesh (which will not have changed)

Hope this helps!

**Reply by sergei9838** [reply]
Thank you, Marco! This is about what I am planning to realise. But it appears that I do not know how vedo treats obj-files. But this will be a separate topic.
